vm/devices/storage/disk_nvme/nvme_driver/src/driver.rs

```diff
@@ -680,10 +936,139 @@ impl<T: DeviceBacking> DriverWorkerTask<T> {
             cpu,
         })
     }
+
+    /// Save NVMe driver state for servicing.
+    pub async fn save(
+        &mut self,
+        worker_state: &mut WorkerState,
+    ) -> anyhow::Result<NvmeDriverSavedState> {
+        let admin = self.admin.as_ref().unwrap().save().await?;
+        let mut io: Vec<QueuePairSavedState> = Vec::new();
+        for io_q in self.io.iter() {
+            io.push(io_q.save().await?);
+        }
+
+        let save_state = NvmeDriverSavedState {
+            admin: Some(admin),
+            io,
+            identify_ctrl: [0; 4096],  // Will be updated by the caller.
+            device_id: "".to_string(), // Will be updated by the caller.
+            namespaces: vec![],        // Will be updated by the caller.
+            qsize: worker_state.qsize,
+            max_io_queues: worker_state.max_io_queues,
+        };
+
+        Ok(save_state)
+    }
 }
 
 impl<T: DeviceBacking> InspectTask<WorkerState> for DriverWorkerTask<T> {
     fn inspect(&self, req: inspect::Request<'_>, state: Option<&WorkerState>) {
         req.respond().merge(self).merge(state);
     }
 }
+
+pub mod save_restore {
+    use super::*;
+
+    /// Save/restore state for NVMe driver.
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct NvmeDriverSavedState {
+        /// Admin queue state.
+        #[mesh(1)]
+        pub admin: Option<QueuePairSavedState>,
+        /// IO queue states.
+        #[mesh(2)]
+        pub io: Vec<QueuePairSavedState>,
+        /// Copy of the controller's IDENTIFY structure.
+        #[mesh(3)]
+        pub identify_ctrl: [u8; 4096],
+        /// Device ID string.
+        #[mesh(4)]
+        pub device_id: String,
+        /// Namespace data.
+        #[mesh(5)]
+        pub namespaces: Vec<crate::namespace::SavedNamespaceData>,
+        /// Queue size as determined by CAP.MQES.
+        #[mesh(6)]
+        pub qsize: u16,
+        /// Max number of IO queue pairs.
+        #[mesh(7)]
+        pub max_io_queues: u16,
+    }
+
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct QueuePairSavedState {
+        #[mesh(1)]
+        /// Which CPU handles requests.
+        pub cpu: u32,
+        #[mesh(2)]
+        /// Interrupt vector (MSI-X)
+        pub msix: u32,
+        #[mesh(3)]
+        pub sq_state: SubmissionQueueSavedState,
+        #[mesh(4)]
+        pub cq_state: CompletionQueueSavedState,
+        #[mesh(5)]
+        pub mem_len: usize,
+        #[mesh(6)]
+        pub pfns: Vec<u64>,
```

Comment:
Didn't I already give this feedback? 1. Why are we saving the PFN list twice, once in the queue pair and once in the SQ/CQ state? 2. Why are we saving a list given that queues are always contiguous? Can't we just save the CQ base and the SQ base?